### PR TITLE
TLS Risks - Certificate Validity Too Long

### PR DESCRIPTION
**Pull request by pacant**
Ticket #1196

### CODE REVIEW

File: src/lib/protocols/tls.c [hunk 1]
```diff
@@ -470,8 +474,13 @@ static void processCertificateElements(struct ndpi_detection_module_struct *ndpi
 		       flow->protos.tls_quic_stun.tls_quic.notAfter, utcDate);
 #endif
 	      }
+	      notAfterYear = utc.tm_year;
+		  notAfterMon = utc.tm_mon;
 	    }
-
```
Comment (lucaderi) [inline src/lib/protocols/tls.c:474]: This check is very complex. In the current code you have https://github.com/ntop/nDPI/blob/dev/src/lib/protocols/tls.c#L476
so it is enough to do something like
if((flow->protos.tls_quic_stun.tls_quic.notAfter-flow->protos.tls_quic_stun.tls_quic.notBefore) > threshold)
 ndpi_set_risk(flow, NDPI_TLS_CERT_VALIDITY_TOO_LONG);

Please rework this code

**Review by lucaderi — COMMENTED** [review]
In the new PR please also add tests/result/tls_certificate_too_long.pcap.out that is used to verify the code

**Comment by IvanNardi** [comment]
Might I suggest an improvement?
Only certificates issued after 01/09/2020 should have a validity not longer than 13 months.
It should be useful to avoid triggering any warnings for certificates issued earlier.
My 2 cents

File: src/lib/protocols/tls.c [hunk 2]
```diff
@@ -454,7 +458,7 @@ static void processCertificateElements(struct ndpi_detection_module_struct *ndpi
 	    for(j=0; j<len; j++) printf("%c", packet->payload[offset+j]);
 	    printf("]\n");
 #endif
-
+
```
Comment (utoni) [inline src/lib/protocols/tls.c:459]: whitespace hell :)

File: src/lib/protocols/tls.c [hunk 3]
```diff
@@ -470,8 +474,13 @@ static void processCertificateElements(struct ndpi_detection_module_struct *ndpi
 		       flow->protos.tls_quic_stun.tls_quic.notAfter, utcDate);
 #endif
 	      }
+	      notAfterYear = utc.tm_year;
+		  notAfterMon = utc.tm_mon;
```
Comment (utoni) [inline src/lib/protocols/tls.c]: confusing indentation

File: src/lib/protocols/tls.c [hunk 4]
```diff
@@ -94,6 +94,8 @@ union ja3_info {
  */
 
 #define NDPI_MAX_TLS_REQUEST_SIZE 10000
+#define TLS_THRESHOLD 34186659 // Threshold for certificate validity
```
Comment (IvanNardi) [inline src/lib/protocols/tls.c:97]: @pacant: where does this value come from?
It seems wrong to me: max validity is 398 day -> 398 * 86400 = 34387200

**Pull request merged**